File: native-wrapper/android-native/GnhBackgroundSync/RemoteNodeSyncGuard.kt
package im.getnowhere.app.backgroundsync

import java.util.concurrent.atomic.AtomicBoolean

/** Process-level guard so background work does not compete with an active sync. */
object RemoteNodeSyncGuard {
    private val inProgress = AtomicBoolean(false)

    fun tryAcquire(): Boolean = inProgress.compareAndSet(false, true)

    fun release() {
        inProgress.set(false)
    }

    fun isInProgress(): Boolean = inProgress.get()
}

File: native-wrapper/ios-native/GnhBackgroundSync/RemoteNodeSyncGuard.swift
import Foundation

/** Process-level guard so background work does not compete with an active sync. */
final class RemoteNodeSyncGuard {
    static let shared = RemoteNodeSyncGuard()
    private let lock = NSLock()
    private var inProgress = false

    private init() {}

    func tryAcquire() -> Bool {
        lock.lock()
        defer { lock.unlock() }
        if inProgress { return false }
        inProgress = true
        return true
    }

    func release() {
        lock.lock()
        inProgress = false
        lock.unlock()
    }

    var isInProgress: Bool {
        lock.lock()
        defer { lock.unlock() }
        return inProgress
    }
}
